File: Java/src/_0617_Merge_Two_Binary_Trees/MyTree.java
package _0617_Merge_Two_Binary_Trees;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Date: 2020/9/4 15:55
 * Content:
 */
public class MyTree {
    private TreeNode root;  // 二叉树的根
    private int size;  // 二叉树节点个数

    public MyTree() {
        root = null;
        size = 0;
    }

    // 根据层次遍历创建
    public MyTree(int[] nums) {
        this();
        if (nums.length == 0 || nums[0] == Integer.MIN_VALUE)
            return;
        root = new TreeNode(nums[0]);
        size++;
        createByLevelOrder(root, nums, 0);
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public TreeNode getRoot() {
        return root;
    }

    // 创建二叉树
    public void createByLevelOrder(int[] nums) {
        if (nums.length == 0 || nums[0] == Integer.MIN_VALUE)
            return;

        root = new TreeNode(nums[0]);
        size++;
        createByLevelOrder(root, nums, 0);
    }

    // 创建以node为根节点的二叉树，index代表当前节点node.val的值为nums[index]
    // 只有nums[index] != Integer.MIN_VALUE 才能进入该函数
    private void createByLevelOrder(TreeNode node, int[] nums, int index) {
        // Integer.MIN_VALUE代表该节点为null
        int n = nums.length;
        if (index >= n)
            return;

        // 0 1 2
        int leftIndex = 2 * index + 1;  // nums[index]对应二叉树节点的左孩子下标为 2 * index + 1
        if (leftIndex <= n - 1) {  // 说明 num[leftIndex] 存在
            if (nums[leftIndex] != Integer.MIN_VALUE) {
                node.left = new TreeNode(nums[leftIndex]);
                size++;
                createByLevelOrder(node.left, nums, leftIndex);
            }
            int rightIndex = leftIndex + 1;  // nums[index]对应二叉树节点的左孩子下标为 2 * index + 2
            if (rightIndex <= n - 1) {  // 只有 num[leftIndex] 存在，考察rightIndex才有意义
                if (nums[rightIndex] != Integer.MIN_VALUE) {
                    node.right = new TreeNode(nums[rightIndex]);
                    size++;
                    createByLevelOrder(node.right, nums, rightIndex);
                }
            }
        }
    }

    // 层次遍历
    public static ArrayList<List<Integer>> levelOrder(TreeNode root) {
        ArrayList<List<Integer>> res = new ArrayList<>();
        if (null == root)
            return null;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int levelNum = 1;  // 当前层的节点数

        while (!queue.isEmpty()) {
            int newLevelNum = 0;
            ArrayList<Integer> level = new ArrayList<>();
            for (int i = 0; i < levelNum; i++) {  // 遍历当前层的所有节点
                TreeNode node = queue.remove();
                level.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                    newLevelNum++;
                }
                if (node.right != null) {
                    queue.add(node.right);
                    newLevelNum++;
                }
            }
            res.add(level);
            levelNum = newLevelNum;
        }
        System.out.println(res);
        return res;
    }

    // 前中后序遍历时空复杂度都一样
    // 时间复杂度：O(n)，n为节点数目
    // 空间复杂度：O(h)，h为二叉树高度
    // 前序遍历：递归算法
    public ArrayList<Integer> preorderTraversal() {
        ArrayList<Integer> res = new ArrayList<>();
        preorderTraversal(root, res);
        return res;
    }
    private void preorderTraversal(TreeNode node, List<Integer> list){
        if (null != node){
            list.add(node.val);
            preorderTraversal(node.left, list);
            preorderTraversal(node.right, list);
        }
    }

    // 中序遍历：递归算法
    public ArrayList<Integer> inorderTraversal() {
        ArrayList<Integer> res = new ArrayList<>();
        inorderTraversal(root, res);
        return res;
    }
    private void inorderTraversal(TreeNode node, List<Integer> list){
        if (null != node){
            inorderTraversal(node.left, list);
            list.add(node.val);
            inorderTraversal(node.right, list);
        }
    }

    // 后序遍历：递归算法
    public ArrayList<Integer> postorderTraversal() {
        ArrayList<Integer> res = new ArrayList<>();
        postorderTraversal(root, res);
        return res;
    }
    private void postorderTraversal(TreeNode node, List<Integer> list){
        if (null != node){
            postorderTraversal(node.left, list);
            postorderTraversal(node.right, list);
            list.add(node.val);
        }
    }
}

File: Python/_0000_study/_auxiliary/mytree/MyTree.py
# coding=utf-8


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


'''
    nu = None
    nums = [
        1,
        2, 2,
        3, nu, 4, 3
    ]
    root = MyTree(nums).getRoot()
'''


class MyTree:
    def __init__(self, nums=[]):
        self.nums = nums
        self.size = 0

        if len(self.nums) == 0 or self.nums[0] == None:
            return
        self.root = TreeNode(self.nums[0])
        self.size += 1
        self._createByLevelOrder(self.root, self.nums, 0)

    def size(self):
        return self.size

    def isEmpty(self):
        return self.size == 0

    def getRoot(self):
        return self.root

    # 层次遍历
    def levelOrder(self, root):
        res = []
        q = []
        if root is not None:
            q.append(root)

        while len(q) > 0:
            level = []
            length = len(q)
            while length > 0:
                t = q[0]
                q.pop(0)
                level.append(t.val)
                if t.left is not None:
                    q.append(t.left)
                if t.right is not None:
                    q.append(t.right)
                length -= 1
            res.append(level)
        return res

    # 前序遍历
    def preorder(self, root):
        res = []
        self._preorder(root, res)
        return res

    # 中序遍历
    def inorder(self, root):
        res = []
        self._inorder(root, res)
        return res

    # 后序遍历
    def postorder(self, root):
        res = []
        self._postorder(root, res)
        return res

    def _createByLevelOrder(self, node, nums, index):
        n = len(nums)
        if index >= n:
            return
        leftIndex = 2 * index + 1
        if leftIndex < n:
            if nums[leftIndex] is not None:
                node.left = TreeNode(nums[leftIndex])
                self.size += 1
                self._createByLevelOrder(node.left, nums, leftIndex)
        rightIndex = leftIndex + 1
        if rightIndex < n:
            if nums[rightIndex] is not None:
                node.right = TreeNode(nums[rightIndex])
                self.size += 1
                self._createByLevelOrder(node.right, nums, rightIndex)

    def _preorder(self, root, res):
        if root is not None:
            res.append(root.val)
            self._preorder(root.left, res)
            self._preorder(root.right, res)

    def _inorder(self, root, res):
        if root is not None:
            self._inorder(root.left, res)
            res.append(root.val)
            self._inorder(root.right, res)

    def _postorder(self, root, res):
        if root is not None:
            self._postorder(root.left, res)
            self._postorder(root.right, res)
            res.append(root.val)

    def __len__(self):
        """返回树中节点数"""
        return self.size

    def __str__(self):  # print该对象使用
        return str(self.levelOrder(self.root))


if __name__ == "__main__":
    nu = None
    nums = [
        1,
        2, 2,
        3, nu, 4, 3
    ]
    root = MyTree(nums).getRoot()
    print(MyTree().levelOrder(root))  # 层次遍历
    print(MyTree().preorder(root))  # 前序遍历
    print(MyTree().inorder(root))  # 中序遍历
    print(MyTree().postorder(root))  # 后序遍历
